src/_P044_P1WifiGateway.ino

```diff
@@ -5,127 +5,223 @@
     P1GatewayServer = new WiFiServer(portnumber);
-    if (nullptr != P1GatewayServer) {
+    if (nullptr != P1GatewayServer && WiFi.isConnected()) {
       P1GatewayServer->begin();
-      addLog(LOG_LEVEL_DEBUG, String(F("P1   : WiFi server started at port ")) + portnumber);
+      if(serverActive(P1GatewayServer)) {
+        addLog(LOG_LEVEL_INFO, String(F("P1   : WiFi server started at port ")) + portnumber);
+      } else {
+        addLog(LOG_LEVEL_ERROR, String(F("P1   : WiFi server start failed at port ")) +
+            portnumber + String(F(", retrying...")));
+      }
+    }
+  }
+
+  void checkServer() {
+    if (!serverActive(P1GatewayServer) && WiFi.isConnected()) {
+      P1GatewayServer->close();
+      P1GatewayServer->begin();
+      if(serverActive(P1GatewayServer)) {
+        addLog(LOG_LEVEL_INFO, F("P1   : WiFi server started"));
+      }
     }
   }
 
   void stopServer() {
-    clearBuffer();
     if (nullptr != P1GatewayServer) {
+      if (P1GatewayClient) P1GatewayClient.stop();
       P1GatewayServer->close();
-      addLog(LOG_LEVEL_DEBUG, F("P1   : WiFi server closed"));
+      addLog(LOG_LEVEL_INFO, F("P1   : WiFi server closed"));
       delete P1GatewayServer;
       P1GatewayServer = nullptr;
     }
   }
 
-  void clearBuffer() {
-    serial_buffer = "";
-    serial_buffer.reserve(P044_BUFFER_SIZE);
-    bytes_read = 0;
+  bool hasClientConnected() {
+    if (P1GatewayServer->hasClient())
+    {
+      if (P1GatewayClient) P1GatewayClient.stop();
+      P1GatewayClient = P1GatewayServer->available();
+      P1GatewayClient.setTimeout(CONTROLLER_CLIENTTIMEOUT_DFLT);
+      addLog(LOG_LEVEL_INFO, F("P1   : Client connected!"));
+    }
+
+    if (P1GatewayClient.connected())
+    {
+      clientConnected = true;
+    }
+    else
+    {
+      if (clientConnected) // there was a client connected before...
+      {
+        clientConnected = false;
+        addLog(LOG_LEVEL_INFO, F("P1   : Client disconnected!"));
+      }
+    }
+    return clientConnected;
   }
 
-  bool isInit() const {
-  	return nullptr != P1GatewayServer;
+  void handleClientIn() {
+    uint8_t net_buf[P044_NETBUF_SIZE];
+    int count = P1GatewayClient.available();
+    if (count > 0)
+    {
+      size_t net_bytes_read;
+      if (count > P044_NETBUF_SIZE)
+        count = P044_NETBUF_SIZE;
+      net_bytes_read = P1GatewayClient.read(net_buf, count);
+      P1EasySerial->write(net_buf, net_bytes_read);
+      P1EasySerial->flush(); // Waits for the transmission of outgoing serial data to complete
+
+      if (count == P044_NETBUF_SIZE) // if we have a full buffer, drop the last position to stuff with string end marker
+      {
+        --count;
+        // and log buffer full situation
+        addLog(LOG_LEVEL_ERROR, F("P1   : Error: network buffer full!"));
+      }
+      net_buf[count] = 0; // before logging as a char array, zero terminate the last position to be safe.
+      char log[P044_NETBUF_SIZE + 40] = {0};
```

Comment:
`net_buf` and this `log` will both be stack allocated.
Better use `String` for this one (and check if the log level is active), as that one will be heap allocated.
The stack isn't that large and allocating 128 bytes twice + some is quite a lot on the stack.

Reply:
Just moved this code without altering it. See remark in ticket about this part of code being superfluous and imo can be removed.

Reply:
Code removed